### PR TITLE
#159965301 Write descriptive registration validation errors

### PR DESCRIPTION
**Pull request by micahoriaso**
#### What does this PR do?
- Write descriptive validation error messages on user registration

#### Description of Task to be completed?
- Give an error message when a user tries to register with an email that has already been used
- Give an error message when a user tries to register with a username that has already been used
- Give an error message when the user tries to register with an invalid email address
- Give an error message when the user tries to register with a password that is less than eight characters long
- Give an error message when the user tries to register with a password that is not alphanumeric

#### How should this be manually tested?
$ git clone https://github.com/andela/ah-magnificent6.git
$ git checkout ft-signup_registration-validation-errors-159965301
$ cd ah-magnificent6
$ cp .env-sample .env [and set the `DATABASE_URL` and `SECRET_KEY`]
$ python3 -m venv venv
$ source venv/bin/activate
(venv) $ pip install -r requirements.txt
(venv) $ python manage.py migrate
(venv) $ python manage.py runserver
Using [Postman](https://www.getpostman.com/) navigate to 

> {your-server-root}:8000/api/users/signup/

with a payload of this format

```
{
"email":"your-email-here",
"username":"your-username",
"password":"your-password"
}
```
for example 

```
{
"email":"test@test.com",
"username":"testuser",
"password":"1235678910"
}
```

#### What are the relevant pivotal tracker stories?
[#159965301](https://www.pivotaltracker.com/story/show/159965301)
#### Screenshots
screen shot 2018-09-05 at 16 12 22
screen shot 2018-09-05 at 16 12 50
screen shot 2018-09-05 at 16 16 03

### CODE REVIEW

**Comment by micahoriaso** [comment]
@PAbishai, @meshack-mbuvi kindly take time to review this Pull Request

**Comment by pabishai** [comment]
@micahoriaso based on your description we will be testing develop

File: authors/apps/authentication/views.py [hunk 1]
```diff
@@ -18,22 +19,21 @@ class RegistrationAPIView(generics.CreateAPIView):
...
     renderer_classes = (UserJSONRenderer,)
     serializer_class = RegistrationSerializer
 
     def post(self, request):
         # Separate requests
-        email, username, password = request.data.get('email', None)\
-                                    , request.data.get('username', None)\
-                                    , request.data.get('password', None)
+        email, username, password = request.data.get('email', None), request.data.get(
+            'username', None), request.data.get('password', None)
```
Comment (pabishai) [inline authors/apps/authentication/views.py]: @micahoriaso is this assignment pep8 standards?

**Comment by meshack-mbuvi** [comment]
Am just curious, do we need to send `username` and `email` to users as a response when they create their accounts? I am in the opinion of just sending to them a welcome message and token.

**Comment by meshack-mbuvi** [comment]
Please add a step to set the environmental variables for the project. Without it, people may not know whether they should have the configurations for the DATABASE and SECRET_KEY.

File: authors/apps/authentication/views.py (hunk 1 above)
Comment (meshack-mbuvi) [reply authors/apps/authentication/views.py]: I think it does not violate pep8 standard @PAbishai

**Comment by micahoriaso** [comment]
@meshack-mbuvi  the environment variables issue has been resolved

**Comment by micahoriaso** [comment]
@PAbishai I have included the feature branch step in the pull request

File: authors/apps/authentication/views.py (hunk 1 above)
Comment (micahoriaso) [reply authors/apps/authentication/views.py]: @PAbishai yes, this follows pep8 standards.

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -16,22 +18,91 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
         fields = ['email', 'username', 'password']
 
+    def validate(self, data):
+        """
+        The `validate` method is where we make sure that the current
+        instance of `RegistrationSerializer` has "valid". In the case of registering
+        a user, this means validating that they've provided an email, username
+        and password.
+        """
+        email = data.get('email', None)
```
Comment (parseen254) [inline authors/apps/authentication/serializers.py]: @micahoriaso This can be compressed to a single line variable assignment. i.e a,b = valueA, valueB

**Review by parseen254 — CHANGES_REQUESTED** [review]

**Comment by pabishai** [comment]
@micahoriaso  It works as expected

File: authors/apps/authentication/tests/test_registration.py [hunk 3]
```diff
@@ -78,3 +91,43 @@ def test_unsuccessful_registration_without_password(self):
...
+    def test_unsuccessful_registration_with_all_numeric_password(self):
+        """ Test that a user enters a password with numbers only """
+        self.valid_user['password'] = "123409275"
+        response = self.client.post(
+            self.registration_url, self.valid_user, format='json')
+        self.assertEqual(response.data['errors']['error'][0],
+                         self.violate_password_pattern_message)
+
+    def test_unsuccessful_registration_with_all_letter_password(self):
+        """ Test that a user enters a password with numbers only """
```
Comment (parseen254) [inline authors/apps/authentication/tests/test_registration.py]: @micahoriaso make sure the descriptions are in sync with the implementation.

**Review by parseen254 — CHANGES_REQUESTED** [review]

**Review by parseen254 — APPROVED** [review]

**Pull request merged**